### ISSUE TITLE
QUASI_RANDOM repeats same classes in one batch (not random?)

### ISSUE THREAD

**Issue by ArneNx** [libffcv/ffcv#304]
I am trying to run some ImageNet training in my own setup. 
Since I am not getting the same result reported in the literature, I am currently investigating what's going wrong. 
I found that when I use `QUASI_RANDOM` as the order option for the FFCV loader, I get non-random composition of batches.
A batch of 16 images has only between 10 and 13 unique classes. This happens consistently and is highly unlikely with 1000 classes. 
Choosing `RANDOM` as the ordering solves this issue, but significantly slows down my training. Is there any explanation for this or is this a bug?
I haven't finished my training yet, but I can already see a significant difference in validation performance between the two orderings.

**Comment by GuillaumeLeclerc** [comment]
As the name imply, QUASI_RANDOM isn't RANDOM. So yes it will not be unbiased, especially if you have small batches like this. The dataset is split in pages when converted to a `.beton`. We currently keep `2*batch_size` in a buffer and sample from those. In the worst case scenario all the samples in a page are from the same class which means for each batch you
pick 16 out of 32 classes with repetition. Add the fact that in the buffer you might have the same class repeated and what you describe is perfectly in line with what one would expect. If you want to solve this, in order of ease of implementation:

- Pre-shuffle your train set. There is no reason it should be sorted by class. If you convert from a pytorch dataset, just add `shuffle=True` to the loader
- Increase the page size. In your `Writer` when you generate the beton you can make the pages bigger, there is a higher chance that you will have multiple classes in a given page
- Increase the size of the buffer (current value=2xbatch_size). This is on the last line of the `QuasiRandomSampler` If i recal properly but it is hard coded so you will have to clone the repo from source. If all above fail and you think this is really necessary and have an idea of how to expose it to end users feel free to submit a PR.

Best!

**Comment by ArneNx** [comment]
Thanks for the explanation.
I thought it'd be something like this. 
Increasing the size of the buffer helps to some extent. I changed it to `max(2* self.loader.batch_size, 512)`, so it will be reliable for a smaller `batch_size`. 

I did not really get what you meant with "just add `shuffle=True` to the loader". The dataset writer expects a `Dataset` class, not a loader, right? However, I added the option `shuffle_indices=True` to `writer.from_indexed_dataset(...)`. That also helped with my issue. 

What I am still confused about is how this is handled in the ImageNet example. [Here](https://github.com/libffcv/ffcv-imagenet/blob/f134cbfff7f590954edc5c24275444b7dd2f57f6/write_imagenet.py) the dataset is not shuffled beforehand. If I do this, even for a batch-size of 1024, I see that a batch has on average (over 100 batches) 557.2 unique elements as opposed to 640.78 when I use `RANDOM` instead. 
Is there something else going on in the example that I'm missing?

**Comment by ArneNx** [comment]
Maybe it would be helpful to document this somewhere (especially since this could become an issue with small batch-sizes)?
I believe most people are interested in a good compromise between accuracy and speed, rather than peak speed.
